src/main/java/com/aws/greengrass/lifecyclemanager/Kernel.java

```diff
@@ -454,48 +495,72 @@ public GreengrassService locate(String name) throws ServiceLoadException {
+                        v -> {
+                            try {
+                                return createGreengrassServiceInstance(
+                                        v, name, this::locateIgnoreError);
+                            } catch (ServiceLoadException e) {
+                                logger.atError().log("Cannot load service", e);
+                                return new UnloadableService(
+                                        config.lookupTopics(
+                                                DEFAULT_VALUE_TIMESTAMP,
+                                                SERVICES_NAMESPACE_TOPIC,
+                                                name),
+                                        e);
+                            }
+                        });
     }
 
-    private void executeBootstrapTasksAndShutdown(BootstrapManager bootstrapManager, Path bootstrapTaskFilePath)
+    private void executeBootstrapTasksAndShutdown(
+            BootstrapManager bootstrapManager, Path bootstrapTaskFilePath)
             throws ServiceUpdateException, IOException {
         int exitCode = bootstrapManager.executeAllBootstrapTasksSequentially(bootstrapTaskFilePath);
         if (!bootstrapManager.hasNext()) {
-            logger.atInfo().log("Completed all bootstrap tasks. Continue to activate deployment changes");
+            logger.atInfo().log(
+                    "Completed all bootstrap tasks. Continue to activate deployment changes");
         }
-        // If exitCode is 0, which happens when all bootstrap tasks are completed, restart in new launch
-        // directories and verify handover is complete. As a result, exit code 0 is treated as 100 here.
-        logger.atInfo().log((exitCode == REQUEST_REBOOT ? "device reboot" : "Nucleus restart")
-                + " requested to complete bootstrap task");
+        // If exitCode is 0, which happens when all bootstrap tasks are completed, restart in new
+        // launch
+        // directories and verify handover is complete. As a result, exit code 0 is treated as 100
+        // here.
+        logger.atInfo().log(
+                (exitCode == REQUEST_REBOOT ? "device reboot" : "Nucleus restart")
+                        + " requested to complete bootstrap task");
 
         shutdown(30, exitCode == REQUEST_REBOOT ? REQUEST_REBOOT : REQUEST_RESTART);
     }
 
-    @SuppressWarnings(
-            {"UseSpecificCatch", "PMD.AvoidCatchingThrowable", "PMD.AvoidDeeplyNestedIfStmts", "PMD.ConfusingTernary"})
-    private GreengrassService createGreengrassServiceInstance(Context.Value v, String name, CrashableFunction<String,
-            GreengrassService, ServiceLoadException> locateFunction) throws ServiceLoadException {
+    @SuppressWarnings({
+        "UseSpecificCatch",
+        "PMD.AvoidCatchingThrowable",
+        "PMD.AvoidDeeplyNestedIfStmts",
+        "PMD.ConfusingTernary"
+    })
+    private GreengrassService createGreengrassServiceInstance(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 608-670 into a separate method.